### ISSUE TITLE
CTCE v2: Problem with VTAM 3

### ISSUE THREAD

**Issue by Peter-J-Jansen** [SDL-Hercules-390/hyperion#284]
I have a CTCE v2 problem that breaks a successfully established and working connection. It occurs frequently enough, sometimes after a few seconds, sometimes only after minutes of successful communication. It only occurs for VTAM non-MPC CTCA links where at least one side is VTAM 3.3 on a VM/SP system. It does not occur on VTAM 4 systems on z/VM. All communication breakups occur following this scenario:

1. Side X sends a control (CTL) command to side Y.     
2. Side Y receives the CTL, generates an device attention (CSW_ATTN) for it which succeeds with RC=0. 
3. Side Y no longer reacts the way it should, i.e. responding to the X side with a sense command byte (SCB) command.  It looks like the attention I/O interrupt never occurred.
4. After 15 seconds, the X side notices the other Y-side not responding and starts a recovery attempt with a selective reset (RST).

A typical Hercules log shows these events on the Y side:

```
12:20:56 HHC05079I 0:0600 CTCE: -> 0:0600 #008C cmd=WRT=01 xy=aa->Wa l=0008 k=7A0311C6 Stat=00 CC=1 w=0,r=0 WAIT ATTN
12:20:56 HHC00979D ctce: 0:0600: +0000> 01030800 8C010102 18000006 01000000  ................ ................
12:20:56 HHC00979D ctce: 0:0600: +0010> 00000001 4A130A7F                    ....J...         ......."        
12:20:56 HHC05079I 0:0600 CTCE: <- 0:0600 #008D cmd=RED=02 xy=Aa<-wa l=7FF0 k=7BF1038E Stat=0C CC=1 w=0,r=0 MATCH
12:20:56 HHC00979D ctce: 0:0600: +0000< 0202F07F 8D010102 10000006 01000000  ................ ..0"............
12:20:56 HHC01302I 0:0600 CHAN: idaw 00E8E010, len 008: =>00000001 4A130A7F                   ......."
12:20:56 HHC01312I 0:0600 CHAN: stat 0C00, count 0000
12:20:56 HHC01315I 0:0600 CHAN: ccw 02DD11F8 24007FF0
12:20:56 HHC05079I 0:0600 CTCE: -> 0:0600 #008D cmd=RED=02 xy=aa->Ra l=7FF0 k=7BF1038E Stat=00 CC=0 w=0,r=0 WAIT ATTN
12:20:56 HHC00979D ctce: 0:0600: +0000> 0202F07F 8D010102 10000006 01000000  ................ ..0"............
12:20:56 HHC05079I 0:0600 CTCE: <- 0:0600 #008E cmd=WRT=01 xy=Aa<-ra l=0022 k=062967C8 Stat=0C CC=0 w=1,r=0 MATCH
12:20:56 HHC00979D ctce: 0:0600: +0000< 01032200 8E010102 32000006 01000000  ..".....2....... ................
12:20:56 HHC00979D ctce: 0:0600: +0010< 001A0001 0A7F4A13 41000000 1000401C  ......J.A.....@. ....."........ .
12:20:56 HHC00979D ctce: 0:0600: +0020< 0000000F 00000003 1D000000 00000000  ................ ................
12:20:56 HHC00979D ctce: 0:0600: +0030< 0000                                 ..               ..              
12:20:56 HHC01312I 0:0600 CHAN: stat 0C00, count 7FCE=>001A0001 0A7F4A13 41000000 1000401C ....."........ .
12:20:56 HHC01316I 0:0600 CHAN: csw 60, stat 0C00, count 7FCE, ccw DD11F8
12:20:56 HHC05079I 0:0600 CTCE: <- 0:0600 #008F cmd=CTL=17 xy=aC<-aa l=0001 k=04000099         CC=0 w=1,r=1 WAIT ATTN
12:20:56 HHC00979D ctce: 0:0600: +0000< 17010100 8F010102 10000006 01000000  ................ ................
12:20:56 HHC01305I 0:0600 CHAN: attention
12:20:56 HHC05079I 0:0600 CTCE: <- 0:0600 #008F cmd=CTL=17 xy=aC<-aa l=0001 k=04000099 Stat=80 CC=0 w=1,r=1 WAIT ATTN->RC=0
12:20:56 HHC00979D ctce: 0:0600: +0000< 17010100 8F010102 10000006 01000000  ................ ................
12:20:56 HHC01316I 0:0600 CHAN: csw 60, stat 8000, count 0000, ccw 000000
12:21:11 HHC05079I 0:0600 CTCE: <- 0:0600 #0090 cmd=RST=00 xy=aN<-ac l=0000 k=04010491              w=1,r=1 SENSE=4100 RESET
12:21:11 HHC00979D ctce: 0:0600: +0000< 00050000 90010102 10000006 01000000  ................ ................
12:21:15 HHC05079I 0:0600 CTCE: <- 0:0600 #0091 cmd=RST=00 xy=an<-an l=0000 k=04010490              w=1,r=1 SENSE=4100 RESET
12:21:15 HHC00979D ctce: 0:0600: +0000< 00050000 91010102 10000006 01000000  ................ ....j...........
12:22:31 HHC01603I * kicked out of VTAM session
12:22:40 HHC01603I log off
HHC02101I Logger: log closed
```

I suspect that the attention I/O interrupt on the Y side never happens.  Apart from the general request for help with this specific S/370-only problem, I would appreciate it to receive pointers as to how I can debug the suspected absence of the attention I/O interrupt on the Y side.

Any help will be much appreciated!

Thanks,

Peter

**Comment by Fish-Git** [comment]
> ... I would appreciate it to receive pointers as to how I can debug the suspected absence of the attention I/O interrupt on the Y side.

There is no "suspected" about it, Peter.

Your log _**proves**_ that the interrupt is indeed _**not occurring.**_

The existence of messages HHC01302I, HHC01315I, HHC01316I, and HHC01305I in your log tells us that ccw I/O tracing is indeed active on device 600 (i.e. `t+600`), and the the <i><b>non-</b>existence</i> of message `HHC00804I` tells us beyond any shadow of doubt that the I/O interrupt is in fact _not occurring._

Another way to prove it to yourself, would be to enable ptt i/o tracing (`'ptt io 100000'`) and then  displaying the ptt trace table (`'ptt'`) within 1-2 seconds after seeing message `HHC01305I 0:0600 CHAN: attention` appear.

If you don't see a `*IOINT` ptt trace entry whose first value is 0000000000000600 (device 600), then you know the interrupt did _not_ occur. Otherwise if you _do_ see such a ptt trace entry, then the interrupt _did_ in fact occur:

https://github.com/SDL-Hercules-390/hyperion/blob/bdf46c225ec154055942a969044594b4fff60ef5/cpu.c#L1081-L1087

But really, the ptt trace would be overkill. The fact that message `HHC00804I` is nowhere to be seen anywhere in your log is proof enough _(to me!)_ that the interrupt is in fact _not_ occurring:

https://github.com/SDL-Hercules-390/hyperion/blob/bdf46c225ec154055942a969044594b4fff60ef5/cpu.c#L1127-L1136

So now the question becomes: _WHY_ isn't the interrupt occurring? And for that question I'm afraid I don't have an answer. &nbsp; `:(`

My only suggestion would be to insert some LOGMSG statements (or custom PTT trace statements) into the channel.c code that is supposed to be queuing the actual interrupt (to verify the interrupt is indeed being queued), as well as in the cpu.c `present_io_interrupt` function to try and determine why it's apparently returning `0` (zero, i.e. "no interrupt was presented").

One place or the other -- i.e. either the logic that is QUEUING of the interrupt or the logic responsible for PRESENTING the interrupt -- is obviously taking a wrong turn somewhere.

**Comment by wably** [comment]
Fish,

You wrote this:

          But really, the ptt trace would be overkill. The fact that message HHC00804I is nowhere to be 
          seen anywhere in your log is proof enough (to me!) that the interrupt is in fact not occurring

However, I'd point out that message HHC00804I is not anywhere in the entire log file, even for the earlier parts of the transmission/connection that were working properly. Thus in this case, I don't think its absence can be conclusive.

Regards,
Bob

**Comment by Fish-Git** [comment]
> Thus in this case, I don't think its absence can be conclusive.

If i/o (ccw) tracing was enabled to entire time, then it is _absolutely **100%** conclusive._

Now it may be true that the fact that it is not occurring is not conclusive that that is what is causing his problem, I don't know. That I can't say. His problem could well be being caused by something else entirely.

_BUT_... one thing is certain: the fact that we don't see message HHC00804I (even though ccw tracing is active) is _absolutely_ conclusive that the interrupt is not occurring. That we know with certainty. Whether that's his problem or not is of course still unknown. But whether the device attention interrupt -- which _was_ _(supposedly!)_ queued -- is actually occurring or not, _is_ known with certainty.

Make of that what you will. &nbsp; _&lt;shrug&gt;_

**Comment by Peter-J-Jansen** [comment]
Additional experiments have shown that with active CCW tracing (in our case `t+600` ), HHC00804I is _never_ seen. (The same was verified to be case in spanhawk for message HHCCP044I.) By commenting the `if (CPU_STEPPING_OR_TRACING(regs, 0))` test on line 1378 in "cpu.c" message HHC00804I was correctly seen, so perhaps we have a bug in the `CPU_STEPPING_OR_TRACING` macro. I will open a different new Issue for this problem.

As to the problem this Issue #284 addresses however, the suspicion that the ATTN I/O interrupt was not happening appears to be wrong. Even when the (intermittent) problem occurs, the presence of HHC00804I proves that :

```
12:46:19 HHC01315I 0:0600 CHAN: ccw 026A81F8 24007FF0
12:46:19 HHC05079I 0:0600 CTCE: <- 0:0600 #008A cmd=WRT=01 xy=aW<-aa l=0008 k=C3083D8E         CC=1 w=0,r=1 WAIT ATTN->NONE
12:46:19 HHC05079I 0:0600 CTCE: -> 0:0600 #0088 cmd=RED=02 xy=aw->aA l=7FF0 k=75F3048B Stat=0C CC=0 w=0,r=0 MATCH
12:46:19 HHC01312I 0:0600 CHAN: stat 0C00, count 7FE8=>00000001 043E03C8 40000000 20008006 .......H .......
12:46:19 HHC01316I 0:0600 CHAN: csw 60, stat 0C00, count 7FE8, ccw 6A81F8
12:46:19 HHC00804I Processor CP00: I/O interrupt code 0:0600 CSW 606A81F8 0C007FE8
12:46:19 HHC05079I 0:0600 CTCE: <- 0:0600 #008B cmd=CTL=17 xy=aC<-aa l=0001 k=0A02019D         CC=0 w=0,r=0 WAIT ATTN
12:46:19 HHC01305I 0:0600 CHAN: attention
12:46:19 HHC05079I 0:0600 CTCE: <- 0:0600 #008B cmd=CTL=17 xy=aC<-aa l=0001 k=0A02019D Stat=80 CC=0 w=0,r=0 WAIT ATTN->RC=0
12:46:19 HHC01316I 0:0600 CHAN: csw 60, stat 8000, count 0000, ccw 000000
12:46:19 HHC00804I Processor CP00: I/O interrupt code 0:0600 CSW 60000000 80000000
12:46:19 HHC01334I 0:0600 CHAN: ORB: IntP:00000000 Key:6 LPM:00 Flags:00000 ............ ........ CCW:006A81E0
12:46:19 HHC01315I 0:0600 CHAN: ccw 146A8171 60000001
12:46:19 HHC05079I 0:0600 CTCE: -> 0:0600 #0089 cmd=SCB=14 xy=ac->aA l=0001 k=0A02049C Stat=0C CC=1 w=0,r=0 MATCH SCB=17=CTL
12:46:19 HHC01312I 0:0600 CHAN: stat 0C00, count 0000=>17                                  .
12:46:19 HHC01315I 0:0600 CHAN: ccw 016A827C 74000008=>00696010 00696800                   ..-.....
12:46:19 HHC05079I 0:0600 CTCE: <- 0:0600 #008C cmd=RED=02 xy=aR<-aa l=7FF0 k=75F3028F         CC=1 w=0,r=0 WAIT ATTN->NONE
12:46:19 HHC05079I 0:0600 CTCE: -> 0:0600 #008A cmd=WRT=01 xy=ar->aA l=0008 k=350F1A8E Stat=0C CC=1 w=0,r=0 MATCH
12:46:19 HHC01302I 0:0600 CHAN: idaw 00696010, len 008: =>00000001 041E043E                   ........
12:46:19 HHC01312I 0:0600 CHAN: stat 0C00, count 0000
12:46:19 HHC01315I 0:0600 CHAN: ccw 026A81F8 24007FF0
12:46:19 HHC05079I 0:0600 CTCE: -> 0:0600 #008B cmd=RED=02 xy=aa->Ra l=7FF0 k=75F30288 Stat=00 CC=0 w=0,r=0 WAIT ATTN
12:46:19 HHC05079I 0:0600 CTCE: <- 0:0600 #008D cmd=WRT=01 xy=Aa<-ra l=0022 k=106527C5 Stat=0C CC=0 w=1,r=0 MATCH
12:46:19 HHC01312I 0:0600 CHAN: stat 0C00, count 7FCE=>001A0001 043E041E 41000000 10004006 .............. .
12:46:19 HHC01316I 0:0600 CHAN: csw 60, stat 0C00, count 7FCE, ccw 6A81F8
12:46:19 HHC00804I Processor CP00: I/O interrupt code 0:0600 CSW 606A81F8 0C007FCE
12:46:19 HHC05079I 0:0600 CTCE: <- 0:0600 #008E cmd=CTL=17 xy=aC<-aa l=0001 k=0A020198         CC=0 w=1,r=1 WAIT ATTN
12:46:19 HHC01305I 0:0600 CHAN: attention
12:46:19 HHC05079I 0:0600 CTCE: <- 0:0600 #008E cmd=CTL=17 xy=aC<-aa l=0001 k=0A020198 Stat=80 CC=0 w=1,r=1 WAIT ATTN->RC=0
12:46:19 HHC01316I 0:0600 CHAN: csw 60, stat 8000, count 0000, ccw 000000
12:46:19 HHC00804I Processor CP00: I/O interrupt code 0:0600 CSW 60000000 80000000
12:46:34 HHC05079I 0:0600 CTCE: <- 0:0600 #008F cmd=RST=00 xy=aN<-ac l=0000 k=0A03058E              w=1,r=1 SENSE=4100 RESET
12:46:34 HHC01329I 0:0600 CHAN: halt I/O
12:46:34 HHC05079I 0:0600 CTCE: -> 0:0600 #008C cmd=RST=00 xy=aa->Aa l=0000 k=0A03058D              w=1,r=1 SENSE=4100 RESET
12:46:34 HHC01334I 0:0600 CHAN: ORB: IntP:00000000 Key:6 LPM:00 Flags:00000 ............ ........ CCW:006A8458
12:46:34 HHC01315I 0:0600 CHAN: ccw 176A8171 60000001
12:46:34 HHC05079I 0:0600 CTCE: -> 0:0600 #008D cmd=CTL=17 xy=aa->Ca l=0001 k=0A02019B Stat=00 CC=1 w=1,r=1 WAIT ATTN SENSE=4100
```

Following the receipt of ConTroL (CTL) command `#008E` an ATTN is correctly generated (RC=0) the 9th log record from the bottom, and two lines below HHC00804I is seen.  And then the receiving side remains idle for exactly 15 seconds to then start recovery with an HIO.  In the same log records shown above one can also observe a correctly replied to CTL command `#008B` : in log record 14 the correct Sense Command Byte (SCB)  `#0089` is seen following HHC00804 3 records above that.

By the way, with the presence of additional debugging code the problem seems to occur a lot less often.  E.g. with following a `ptt io 100000` command I was once unable to provoke the error.  (Also under spinhawk I have been unable to reproduce the error.)  I do not know if this is significant or not.  This area of Hercules code is quite involved, I only understand fragments of it.  But that is of course my problem.

So the debugging continues for this VTAM 3 only problem ...

Cheers,

Peter

**Comment by Fish-Git** [comment]
Peter Jansen wrote:

> By commenting the `if (CPU_STEPPING_OR_TRACING(regs, 0))` test on line 1378 in "cpu.c" message HHC00804I was correctly seen, so perhaps we have a bug in the `CPU_STEPPING_OR_TRACING` macro.

(Ack!) 

&nbsp;
Bob Polmanter wrote:
> Fish wrote:
> > Thus in this case, I don't think its absence can be conclusive.
> 
> If i/o (ccw) tracing was enabled the entire time, then it is _absolutely **100%** conclusive._

It looks like ONCE AGAIN that I am going to have to eat crow. I owe you both an apology (and a big THANK YOU to Peter for identifying my error, i.e. the flaw in my argument).

As Bob correctly stated (you were right Bob, and I was wrong!), the absence of message `HHC00804I` is indeed _**inconclusive**_, not conclusive as I was erroneously arguing.

As Peter has made quite clear (and as Bob previously eluded to), the flaw in my reasoning was in my presuming that the `if (CPU_STEPPING_OR_TRACING(regs, 0))` statement in "cpu.c" (which in my version of SDL Hyperion is on line [**1128**](https://github.com/SDL-Hercules-390/hyperion/blob/bdf46c225ec154055942a969044594b4fff60ef5/cpu.c#L1127-L1136), not 1378) would evaluate to "true" if ccw tracing was active.

_This of course is **wrong!**_

(and where the flaw in my reasoning lies)

If you examine the [**`CPU_STEPPING_OR_TRACING`**](https://github.com/SDL-Hercules-390/hyperion/blob/bdf46c225ec154055942a969044594b4fff60ef5/opcode.h#L475-L478) macro, you will see that it is checking the `regs->tracing` flag, which is the _**instruction**_ tracing flag, _not_ the _**device**_ (ccw) tracing flag!

The _device_ (ccw) tracing flag is in **`DEVBLK`**, not `REGS`! &nbsp; _(Doh!)_

So message `HHC00804I` is only issued when _instruction_ tracing is active (`t+`), but not when _device_ tracing is active! (`t+devnum`)

My bad. &nbsp; `:(`

> I will open a different new Issue for this problem.

Please do!  Thank you.  As we are all now discovering, such a seemingly small thing can end up causing a lot of confusion, making debugging other problems more difficult. So even though it might seem like a small thing (and on the grand scheme of things it is a small thing), it is nevertheless an _**important** thing!_

So yes, please report it so we can it fixed. Thanks.

The fix, I believe, would be to add `CCW_TRACE_OR_STEP` macro test to the condition (i.e. `if (CPU_STEPPING_OR_TRACING(regs, 0) || CCW_TRACE_OR_STEP(dev))`. But doing so is not going to be so simple however. The channel.c `present_io_interrupt` function is going to need to change to pass back a pointer to the DEVBLK the interrupt is for, so that cpu.c 's `perform_io_interrupt` function can then use it. (Note: there's also another I/O interrupt tracing conditional test 24 lines further down too, that issues message `HHC00805` or `HHC00806` that will also need fixed the same way.)

Anyway, I just wanted to apologize to both Peter and especially to Bob, and to congratulate you both for being on your toes. When I argue something I do so with passion, but sometimes my passion gets the better of me. My mind saw the `CPU_STEPPING_OR_TRACING` statement as a check for ccw tracing because ccw tracing is what I was thinking about at the time.

As all software developers know, sometimes you can be staring a bug right in the face and not see it. Such was the case here. I'll try to be more careful.

**Comment by Fish-Git** [comment]
> By the way, with the presence of additional debugging code the problem seems to occur a lot less often. E.g. with following a `ptt io 100000` command I was once unable to provoke the error. (Also under spinhawk I have been unable to reproduce the error.) I do not know if this is significant or not.

Sounds like a race condition somewhere to me.  If your code is designed to always issue a SCB whenever an ATTN interrupt occurs but doesn't always do so, then something is obviously changing one or more of the conditions that your "SCB" code is testing for. And that "something" is almost always some other thread.

You're probably updating the same flag in two different threads somewhere which obviously needs to be done under control of a lock. Maybe you added a new flag recently and forgot to obtain some lock before updating it. _&nbsp;&lt;shrug&gt;_ &nbsp; It's probably something silly like that.

Anyway... good luck in your bug hunt, Peter! I wish I could be more help, but as we can see all I'm doing is confusing the issue so I'll go away and leave you two alone now. Besides, I need to get back to TXF anyway.

**Comment by Peter-J-Jansen** [comment]
> Sounds like a race condition somewhere to me. If your code is designed to always issue a SCB whenever an ATTN interrupt occurs but doesn't always do so [...]

Actually, the "SCB" command which is the normal response to an ATTN being rasied after receiving a "CTL" command is something the O.S. running under Hercules issues. Both VTAM CTCA links and RSCS NJE links do that, at least when all is well. Using Hercules PTT and extra log messages debugging, the ATTN appears to be correctly signaled. Using Fish's example in "channel.c", I do see the ATTN I/O Interrupt correctly being _queued_ **and** _dequeued_, and in the successful cases followed by a Test CHannel (TCH) and Start I/O (SIO), the latter of which issues the "SCB" CCW command, When the failure occurs, there is no sign of the TCH + SIO commands. Why that is so, I do not know.

As this problem only occurs for VM/SP, I suspected a S/370-only problem. So I decided to backport my CTCE code to spinhawk to see how it behaved under spinhawk. And indeed, under spinhawk, my same CTCE v2 code does _not_ experience this problem, at least I have not been able to reproduce it for VTAM CTCA links un der VM/SP. Also for RSCS NJE links using CTCE v2 in VM/SP under spinhawk, the problem does not appear. Following my very last test earlier today I should now probably say _did not appear most of the time_. Once, _just once_, did I experience a RSCS NJE timeout following a "CTL" which, correctly received at the other end, was not responded to by a "SCB" command.

I've been re-re-reading chapter 13 (and following chapters) in both S/370 and recent PoP's, but I'm close to having admit defeat here. After an apparent successful ATTN interrupt being _queued_ **and** _dequeued_, I simply do not know what to do in order to find out why the TCH + SIO that should follow, actually do not come.

In short, I think I need help from the real experts in this area !

Thanks in advance,

Peter

**Comment by mcisho** [comment]
Hello Peter,

To me it seems you have to determine whether the Attention is being presented to the MVS(?) virtual machine that is running VTAM. Until we know that there are too many areas (Hercules, VM, MVS, or VTAM) that might be playing a part.

The **CP TRACE IO** command could help you determine whether the Attention is being presented to the MVS(?) virtual machine.

The manual says about CTL:
> ... the adapter generates attention status for presentation to the y  channel.
> When  status  is  already  pending  on  the  y side, the adapter holds the
> attention status and does not attempt to  present  it  until  the  pending
> status  has  been cleared.  When an independent command is active on the y
> side, the adapter holds the attention  status  and  does  not  attempt  to
> present  it  until  ending  status for that command has been cleared.
So does Hercules think there is pending status? Off hand I don't know the code well enough to even suggest where to start looking, but if Hercules has such an indicator, one of your existing messages could display yes or no. Your trace doesn't suggest there should be anything pending, but...

If Hercules isn't presenting the Attention, then lots of PTT and/or messages will be needed to work out why not.

Cheers, Ian

p.s. Several years ago I started writing the equivalent of CTCE. I'd got the CCW handling and the socket code working (the simple bits!), but lack of time prevented me making any progress with the hard part. Then CTCE came along and saved me a whole bunch of effort! I would still like to write a CMS program to drive a CTC device, but again lack of time...

**Comment by mcisho** [comment]
Peter, your trace (line 11) has the following message:

> HHC00804I Processor CP00: I/O interrupt code 0:0600 CSW 60000000 80000000

Why does the CSW have the storage key of 6, when *Bits 0-3 form the storage protection
key used in the chain of operations initiated by the last START I/O*, and an Attention *is not associated with the initiation, execution, or termination of any I/O operation*, Quotes from the S/360 PoP. Looks like a Hercules bug to me, though whether it has any impact on this issue is questionable.

Another question, though probably one only the authors of this code can answer. Why are CTL requests being issued? Are they a normal part of the connection? They seem like a completely superfluous overhead to me, providing absolutely no benefit.

**Comment by wably** [comment]
Hello Ian,

I have been working with Peter on this issue and I supplied some of the traces you saw above until Peter was able to recreate the issue himself.  But while Peter is the author of CTCE and the definitive source for this issue, I think I can answer some of your questions.

> To me it seems you have to determine whether the Attention is being presented to the MVS(?) virtual machine that is running VTAM. Until we know that there are too many areas (Hercules, VM, MVS, or VTAM) that might be playing a part.

The OS is VM/SP 5, and the virtual machine using the CTC device is VTAM.  VTAM is connected with another VTAM also under VM/SP 5 on a completely different host computer.  The two VTAMs are connected via CTC, via the CTCE device, in a cross-domain session.

Peter and I determined a couple of weeks ago that the attention interruption was not being presented to the VTAM virtual machine.  That is, tracing at the virtual machine level would show that for the last i/o issued prior to the problem occurring, the attention was generated in the channel (in Hercules) but the virtual machine never received it.  At the time, whether Hercules ever presented that particular interruption to the guest OS, or whether the guest OS did receive it but somehow did not pass it on to the virtual machine remained an open question.

Peter can jump in if I am mistaken but my recollection is that we decided that it just wasn't plausible that the guest OS would fail to present the interruption if it received it; after all, the CTC device is a dedicated device to the VTAM virtual machine.  Hence the assumption was that Hercules was not presenting it, in whatever special circumstance is causing this.  Because up until the failing moment, all the interruptions **are** presented and processed by each component as expected.  Further, and as Peter indicated above, this missing interruption does not occur using Spinhawk with all other components and connections the same, so if it was a guest OS bug it should still happen there as well.

> Why does the CSW have the storage key of 6, when Bits 0-3 form the storage protection
key used in the chain of operations initiated by the last START I/O, and an Attention is not associated with the initiation, execution, or termination of any I/O operation, Quotes from the S/360 PoP. Looks like a Hercules bug to me, though whether it has any impact on this issue is questionable.

VTAM runs in key 6.   The key in the CSW is normal.

> Another question, though probably one only the authors of this code can answer. Why are CTL requests being issued? Are they a normal part of the connection? They seem like a completely superfluous overhead to me, providing absolutely no benefit.

The CTL requests are being issued by VTAM in the normal course of its communications.  I can't speak to the reason they are needed or to the protocols and internals within VTAM that uses these commands and their results.

I hope some of these answers will clarify things until Peter can weigh in.

Regards,
Bob

**Comment by Peter-J-Jansen** [comment]
Ian and Bob, thanks for your assistance with this.

Sorry for not responding any sooner.  I first updated my SDL-Hyperion in order to profit from Fish's latest "I/O interrupt tracing tweak".  And then I got into trouble with the latest VS2109 update (16.5.x), having to disable warning C4789.

The CTC adapter holding the attention status until a pending / ending status has been cleared, I believe is being adhered to by the CTCE driver.  The "device_attention" routine in "channel.c" in that case returns RC==1, and the CTCE driver will then go into an increasingly longer waiting loop to keep on retrying.  A log msg shows this when it happens (busy_waits > 0).  When the ATTN is finally signaled, a successful Queue and Dequeue of the ATTN I/O Interrupt is always being shown using (an adtapted version of) Fish's extra log msgs, but only when all is fine is this being followed by a TCH + SIO as shown by PTT and extra log msgs I added. I presume the LPSW of the I/O Interrupt takes place, but no TCH / SIO follows when the problem as reported occurs.

A correct working sequence as it appears many times :

```
14:16:54 HHC05079I 0:0600 CTCE: -> 0:0600 #008E cmd=RED=02 xy=aa->Ra l=7FF0 k=7EA2797D Stat=00 CC=0 w=1,r=1 WAIT ATTN
14:16:54 HHC05079I 0:0600 CTCE: <- 0:0600 #0090 cmd=WRT=01 xy=Aa<-ra l=0008 k=4B025AB8 Stat=0C CC=0 w=0,r=1 MATCH
14:16:54 HHC91336I 0:0600 CHAN: Normal interrupt Queued; Normal=1, PCI=0, ATTN=0 @ channel.c:522
14:16:54 HHC81336I 0:0600 CHAN: Normal interrupt DEQueued; Normal=0, PCI=0, ATTN=0 @ channel.c:6012
14:16:54 HHC81336I 0:0600 CHAN: Normal interrupt NOT DEQueued; Normal=0, PCI=0, ATTN=0 @ channel.c:1549
14:16:54 HHC90804I Processor CP00: I/O interrupt code 0:0600 CSW 609FE1F8 0C007FE8 icode=-1 load_psw RC=0 PSW=000C00000001F2C0
14:16:54 HHC05079I 0:0600 CTCE: <- 0:0600 #0091 cmd=CTL=17 xy=aC<-aa l=0001 k=7EA20586         CC=0 w=0,r=0 WAIT ATTN
14:16:54 HHC91336I 0:0600 CHAN: ATTN interrupt Queued; Normal=0, PCI=0, ATTN=1 @ channel.c:3948
14:16:54 HHC05079I 0:0600 CTCE: <- 0:0600 #0091 cmd=CTL=17 xy=aC<-aa l=0001 k=7EA20586 Stat=80 CC=0 w=0,r=0 WAIT ATTN->RC=0
14:16:54 HHC81336I 0:0600 CHAN: ATTN interrupt DEQueued; Normal=0, PCI=0, ATTN=0 @ channel.c:6012
14:16:54 HHC81336I 0:0600 CHAN: ATTN interrupt NOT DEQueued; Normal=0, PCI=0, ATTN=0 @ channel.c:1549
14:16:54 HHC90804I Processor CP00: I/O interrupt code 0:0600 CSW 60000000 80000000 icode=-1 load_psw RC=0 PSW=000C00000001F2C0
14:16:54 HHC81335I 0:0600 CHAN: test channel TCH: cc=0
14:16:54 HHC71336I 0:0600 CHAN: start io SIO: Pending Normal=0 PCI=0 ATTN=0 tsch=0 (CC=1); busy=0 startpending=0 (CC=2)
14:16:54 HHC05079I 0:0600 CTCE: -> 0:0600 #008F cmd=SCB=14 xy=ac->aA l=0001 k=7EA2009B Stat=0C CC=1 w=0,r=0 MATCH SCB=17=CTL
14:16:54 HHC05079I 0:0600 CTCE: -> 0:0600 #0090 cmd=WRT=01 xy=aa->Wa l=0008 k=2282D033 Stat=00 CC=1 w=0,r=0 WAIT ATTN
14:16:54 HHC05079I 0:0600 CTCE: <- 0:0600 #0092 cmd=RED=02 xy=Aa<-wa l=7FF0 k=7EA27961 Stat=0C CC=1 w=0,r=0 MATCH
```

And a sequence of commands showing the error :

```
14:31:23 HHC05079I 0:0600 CTCE: <- 0:0600 #000A cmd=WRT=01 xy=aW<-aa l=03E2 k=F970669F         CC=1 w=1,r=0 WAIT ATTN->NONE
14:31:23 HHC05079I 0:0600 CTCE: -> 0:0600 #0001 cmd=RED=02 xy=aw->aA l=7FF0 k=7EA273F2 Stat=0C CC=0 w=1,r=1 MATCH
14:31:23 HHC91336I 0:0600 CHAN: Normal interrupt Queued; Normal=1, PCI=0, ATTN=0 @ channel.c:522
14:31:23 HHC81336I 0:0600 CHAN: Normal interrupt DEQueued; Normal=0, PCI=0, ATTN=0 @ channel.c:6012
14:31:23 HHC81336I 0:0600 CHAN: Normal interrupt NOT DEQueued; Normal=0, PCI=0, ATTN=0 @ channel.c:1549
14:31:23 HHC90804I Processor CP00: I/O interrupt code 0:0600 CSW 6069F1F8 0C007C0E icode=-1 load_psw RC=0 PSW=000C00000001F2C0
14:31:23 HHC05079I 0:0600 CTCE: <- 0:0600 #000B cmd=CTL=17 xy=aC<-aa l=0001 k=7EA2091C         CC=0 w=1,r=1 WAIT ATTN
14:31:23 HHC91336I 0:0600 CHAN: ATTN interrupt Queued; Normal=0, PCI=0, ATTN=1 @ channel.c:3948
14:31:23 HHC05079I 0:0600 CTCE: <- 0:0600 #000B cmd=CTL=17 xy=aC<-aa l=0001 k=7EA2091C Stat=80 CC=0 w=1,r=1 WAIT ATTN->RC=0
14:31:23 HHC81336I 0:0600 CHAN: ATTN interrupt DEQueued; Normal=0, PCI=0, ATTN=0 @ channel.c:6012
14:31:23 HHC81336I 0:0600 CHAN: ATTN interrupt NOT DEQueued; Normal=0, PCI=0, ATTN=0 @ channel.c:1549
14:31:23 HHC90804I Processor CP00: I/O interrupt code 0:0600 CSW 60000000 80000000 icode=-1 load_psw RC=0 PSW=000C00000001F2C0
14:31:41 HHC05079I 0:0600 CTCE: <- 0:0600 #000C cmd=RST=00 xy=aN<-ac l=0000 k=7EA20D0D              w=1,r=1 SENSE=4100 RESET
14:31:41 HHC05079I 0:0600 CTCE: -> 0:0600 #0002 cmd=RST=00 xy=aa->Aa l=0000 k=7EA20D03              w=1,r=1 SENSE=4100 RESET
14:31:41 HHC81335I 0:0600 CHAN: test channel TCH: cc=0
14:31:41 HHC71336I 0:0600 CHAN: start io SIO: Pending Normal=0 PCI=0 ATTN=0 tsch=0 (CC=1); busy=0 startpending=0 (CC=2)
14:31:41 HHC05079I 0:0600 CTCE: -> 0:0600 #0003 cmd=CTL=17 xy=aa->Ca l=0001 k=7EA20914 Stat=00 CC=1 w=1,r=1 WAIT ATTN SENSE=4100
```

The HHC90804I messages is a modified version of HHC00804I, produced within "perform_io_interrupt" in "cpu.c", right after "load_psw" and shows its Return Code == 0.  After the CTL causing the ATTN, this is the last activity seen for 8 seconds.  The TCH plus SIO that normally follow do not show up.

What I still would like to understand better, is the relations between these 3 important locks: the device lock, the io interrupt quueue lock, and the interrupt lock.  The first two of these I think I (somewhat) understand.  The device lock protects the atomicity requirements for device block updates (each device has one), and the interrupt queue lock protects the (doubly linked) queue of io blocks, anchored in sysblk.iointqlk.  But what exactly protects the all-important interrupt lock, i.e. sysblk.intlock ?

The other thing which is still a black art to me is the SIE support for I/O interrupts, and under which conditions these need to be intercepted.  I suspect this may be related to this problem as the CTCA devices are being used by VM / SP guests, VTAM (v3) as well as RSCS (also v3).  Perhaps Ivan can help me here ?

Thanks again to you all for your help with this problem of mine which I have so many difficulties with.

Cheers, and stay healthy !

Peter

**Comment by Fish-Git** [comment]
> But what exactly protects the all-important interrupt lock, i.e. sysblk.intlock?

Say what? You're asking what protects a _lock?_ That doesn't make any sense!

**Comment by Peter-J-Jansen** [comment]
What I meant to ask about the Hercules interrupt lock (sysblk.intlock) is which actions require to be surrounded by an OBTAIN_INTLOCK() / RELEASE_INTLOCK().  I presume that whenever changes are to be made to the interrupt queue (anchored in sysblk.iointqlk), which requires the interrupt queue lock, then the INTLOCK() is required even prior to that, but what else is there that requires that INTLOCK() ? Am I correct to assume it includes all `..._IC_...` actions, which also means all changes to CPU running / waiting / syncing / etc. states ? Or is there even more ?

Thanks,

Peter

**Comment by Fish-Git** [comment]
* Bringing a CPU online/offline
* Change in a CPU's state (start, stop, etc)
* Change in a CPU's architecture mode
* CPU entering/leaving SIE mode
* CPU synchronization (tlb, etc)
* I/O interrupt queue
* ..._IC_...
* Misc: certain sysblk flags (tracing, etc)

There might be more, but that's all I could find after doing a _very quick_ search.

Basically anything to do with the I/O subsystem or the CPUs themselves. Any type of synchronization needing to be done related to architectural emulation. It's pretty much considered to be our "master" lock and is used for just about everything whenever there isn't already a specific lock (and even then intlock usually needs to be acquired beforehand, such as in the I/O interrupt queue lock case).

And of course the occasional need to sometimes acquire (or continue to hold) the device lock too during such situations throws a huge monkey wrench into it all.

It's a mess.

**Comment by Peter-J-Jansen** [comment]
Thanks for the clarifications Fish ! 
And I do agree with your feelings, it 's a bit of mess. I was suspecting a subtile race condition related to this Issue #284, but thanks to help of Bob Polmanter, a fix of a different kind will be committed soon.

Cheers,

Peter

**Comment by wably** [comment]
This issue is being caused by a VM/SP bug (not VTAM).  There is an interruption that is not being presented and as implausible as it seemed before, we now have documented evidence that this is occurring.  Somehow and intermittently, CP receives two interruptions from the CTCA device before the first one can be presented to the virtual machine using the CTCA,  When an interruption is finally presented to the virtual machine (VTAM in this case), the second interruption is what is presented.

The two interruptions that are coming in are normal and correct.  The other side of the CTCE connection issues two SIOs, one a normal write/read CCW sequence (completed normally), and then a CTL CCW command.

When things are working correctly, here is what is happening on the side of the CTCE that is eventually experiencing the problem as a result of those two SIO operations.  I'm showing only the most relevant trace entries and have edited and annotated them for clarity.

> HHC01316I 0:0600 CHAN csw 60, stat 0C00 ....  (Hercules presents CE+DE interruption)
> CP:  CSW ... R 0600 60F92F80 0C000000    (CP receives real interrupt; queues it to VTAM)
> VTAM:  I/O  0600 ...  CSW 0C00    (VTAM receives the virtual interruption)
> HHC01316I 0:0600 CHAN csw 60, stat 8000 ....  (Hercules presents ATTN interruption)
> CP:  CSW ... R 0600 60000000 80000000    (CP receives real interrupt; queues it to VTAM)
> VTAM:  I/O  0600 ...  CSW 8000    (VTAM receives the virtual ATTN interruption)

Here is what is happening when things go awry:

> HHC01316I 0:0600 CHAN csw 60, stat 0C00 ....  (Hercules presents CE+DE interruption)
> CP:  CSW ... R 0600 60F92F80 0C000000    (CP receives real interrupt; queues it to VTAM)
> HHC01316I 0:0600 CHAN csw 60, stat 8000 ....  (Hercules presents ATTN interruption)
> CP:  CSW ... R 0600 60000000 80000000    (CP receives real interrupt; queues it to VTAM)
> VTAM:  I/O  0600 ...  CSW 8000    (VTAM receives the virtual ATTN interruption)

As you can see, the virtual interruption containing the CE+DE X'0C' status was never presented to VTAM.  As a result, VTAM never receives completion of its i/o operation to
read the data sent by the write from the other side.  After 15 seconds, VTAM issues the CTC timeout messages and the connection goes into retry/recovery which never succeeds.  The ATTN that VTAM receives means nothing; it is out of sync.  The VTAM on the other side of the CTCE also times out and enters retry/recovery because it did not receive its response to the CTL CCW it last issued.

A CP solution to this problem is what is needed but it is not apparent at this time how likely I will be able to track it down.  Is this an issue with CP's handling of the real interruption?  Did CP queue it incorrectly (out of order) or overlay the second interruption over top of the first?  If CP handed it off correctly to the virtual i/o interrupt simulation part of CP, did that code queue the interruptions correctly?  Did it overlay one over the other as the interrupt information was placed into the virtual device blocks?  When the CP dispatcher _dequeued_ the interruptions for presentation to the virtual machine, was there an error in this process?

Whatever the case, Peter and I had a discussion about a possible work around solution.  I'll let Peter detail what was decided and how he implemented it when he closes this issue along with his commit.

Regards,
Bob

**Comment by Fish-Git** [comment]
> This issue is being caused by a VM/SP bug (not VTAM). There is an interruption that is not being presented and as implausible as it seemed before, we now have documented evidence that this is occurring.

> A CP solution to this problem is what is needed but it is not apparent at this time how likely I will be able to track it down.

> Whatever the case, Peter and I had a discussion about a possible work around solution. I'll let Peter detail what was decided and how he implemented it when he closes this issue along with his commit.

_**EXCELLENT WORK,**_ you two! &nbsp; `:-D`

Our Hercules development team is _the **best!**_ &nbsp; `:-D`

**Comment by Peter-J-Jansen** [comment]
Commit 5080ee8 has addressed this Issue #284 by implementing an additional option to the CTCE configuration : `ATTNDELAY <delay>` which will delay ATTN signals with `<delay>` msec. The default is 0. VM/SP CTCA links experiencing SIO timeout errors have been circumvented by specifying `ATTNDELAY 200` or so.

With many thanks to Bob Polmanter whose support for this was substantial !

I think this issue can now be closed.

Cheers,

Peter
